smt.go

```diff
@@ -377,6 +377,106 @@ func (smt *SMT) Prove(key []byte) (proof *SparseMerkleProof, err error) {
 	return proof, nil
 }
 
+// ProveClosest generates a SparseMerkleProof of inclusion for the key
+// with the most common bits as the path provided
+func (smt *SMT) ProveClosest(path []byte) (
+	closestPath, closestValueHash []byte, // the closest leaf info for the key provided
+	proof *SparseMerkleProof, // proof of the key-value pair found
+	err error, // the error value encountered
+) {
+	var siblings []treeNode
+	var sib treeNode
+	var parent treeNode
+	var parentDepth int
+
+	node := smt.tree
+	depth := 0
+	for depth < smt.depth() {
+		node, err = smt.resolveLazy(node)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		if node == nil && depth == 0 {
+			break // DISCUSS_IN_THIS_PR: I dont think this is even possible
+		}
+		parentDepth = depth
+		parent = node
+		if node == nil {
+			parent, err = smt.resolveLazy(parent)
+			if err != nil {
+				return nil, nil, nil, err
+			}
+			inner, ok := parent.(*innerNode)
+			if !ok {
+				panic("parent not inner node")
+			}
+			if len(siblings) > 0 {
+				siblings = siblings[:len(siblings)-1]
```

Comment:
What are we trimming here?

Reply:
Added a comment, but if we hit this if statement we have hit a placeholder, prior to hitting it we added the sibling. This sibling is where we want to goto and thus should not be included in the side nodes as it would create an incorrect root hash. Ultimately this is just part of the backstepping.

Reply:
Please add that explanation in the comments. Very valuable information!